### PR TITLE
#3701 fix some pipeline settingDialog defined bugs

### PR DESCRIPTION
**Pull request by wu87988622**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #3701 )
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/context/pipeline/pipelineActions.js [hunk 1]
```diff
@@ -61,32 +53,30 @@ export const createActions = context => {
       if (state.isFetching) return;
       try {
         dispatch(routine.request());
-        // get all endpoints from current pipeline
-        const currentEndpoints = get(currentPipeline, 'endpoints', []);
+
         // we need to merge all endpoints in order to update object
         // group value is decided by component kind
```
Comment (konekoya) [inline ohara-manager/client/src/context/pipeline/pipelineActions.js]: We should probably update the comment as well :)

File: ohara-manager/client/src/context/pipeline/pipelineActions.js [hunk 2]
```diff
@@ -61,32 +53,30 @@ export const createActions = context => {
...
+              break;
+          }
+          return { ...endpoint, group };
+        });
+
         const data = await pipelineApi.update({
           ...values,
-          endpoints: mergedEndpoints,
+          endpoints:
+            groupEndpoints && values.endpoints ? groupEndpoints : undefined,
```
Comment (konekoya) [inline ohara-manager/client/src/context/pipeline/pipelineActions.js]: The `undefined` could be assigned as a group for an object?

File: ohara-manager/client/src/context/pipeline/pipelineActions.js [hunk 3]
```diff
@@ -61,32 +53,30 @@ export const createActions = context => {
...
-              case KIND.topic:
-                group = topicGroup;
-                break;
-              default:
-                break;
-            }
-            return { ...endpoint, group };
-          }),
-        );
+        const groupEndpoints = map(get(values, 'endpoints'), endpoint => {
```
Comment (konekoya) [inline ohara-manager/client/src/context/pipeline/pipelineActions.js]: newEndpoints or updatedEndpoints sounds better to me :)

File: ohara-manager/client/src/components/common/Definitions/Tags.js [hunk 4]
```diff
@@ -28,7 +28,18 @@ const Tags = props => {
     ...rest
   } = omit(props, ['tableKeys']);
 
-  const hasError = (meta.error && meta.touched) || (meta.error && meta.dirty);
+  let hasError = (meta.error && meta.touched) || (meta.error && meta.dirty);
```
Comment (konekoya) [inline ohara-manager/client/src/components/common/Definitions/Tags.js]: I've changed the error displaying timing in #3698, should we also update here?

File: ohara-manager/client/src/components/Pipeline/Graph/Topic/TopicGraph.js [hunk 5]
```diff
@@ -95,14 +97,24 @@ const TopicGraph = params => {
...
+
       this.$box.find('#topic-remove').on('mousedown', async function() {
         await stopTopic(name);
         await deleteTopic(name);
-      });
 
-      this.$box
-        .find('#topic-remove')
-        .on('click', _.bind(this.model.remove, this.model));
+        const removeEndpoints = currentPipeline.endpoints.filter(
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Graph/Topic/TopicGraph.js]: removedEndpoints

File: ohara-manager/client/src/components/Pipeline/Graph/Connector/ConnectorGraph.js [hunk 6]
```diff
@@ -181,14 +184,24 @@ const ConnectorGraph = params => {
...
-      });
-
       this.$box
         .find('#connector-remove')
         .on('click', _.bind(this.model.remove, this.model));
 
+      this.$box.find('#connector-remove').on('mousedown', async () => {
+        await deleteConnector(name);
+
+        const removeEndpoints = currentPipeline.endpoints.filter(
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Graph/Connector/ConnectorGraph.js]: removedEndpoints

File: ohara-manager/client/src/components/Pipeline/Graph/Connector/ConnectorGraph.js [hunk 7]
```diff
@@ -72,7 +74,8 @@ const ConnectorGraph = params => {
   const settingIcon = renderToString(<BuildIcon viewBox="-4 -5 32 32" />);
   const removeIcon = renderToString(<CancelIcon viewBox="-4 -5 32 32" />);
 
-  const isMetricsDisplayed = metrics.meters.length > 0 && isMetricsOn;
+  const isMetricsDisplayed =
+    _.get(metrics, 'meters', []).length > 0 && isMetricsOn;
```
Comment (konekoya) [inline ohara-manager/client/src/components/Pipeline/Graph/Connector/ConnectorGraph.js]: Thanks. Does this fix the issue?

File: ohara-manager/client/src/context/pipeline/pipelineActions.js (hunk 2 above)
Comment (eechih) [reply ohara-manager/client/src/context/pipeline/pipelineActions.js]: 建議寫法 `endpoints: values.endpoints && groupEndpoints` :)

File: ohara-manager/client/src/context/pipeline/pipelineActions.js (hunk 1 above)
Comment (wu87988622) [reply ohara-manager/client/src/context/pipeline/pipelineActions.js]: ok

File: ohara-manager/client/src/components/common/Definitions/Tags.js (hunk 4 above)
Comment (wu87988622) [reply ohara-manager/client/src/components/common/Definitions/Tags.js]: this change is only by tags

File: ohara-manager/client/src/context/pipeline/pipelineActions.js (hunk 2 above)
Comment (wu87988622) [reply ohara-manager/client/src/context/pipeline/pipelineActions.js]: @eechih thanks~

File: ohara-manager/client/src/components/Pipeline/Graph/Connector/ConnectorGraph.js (hunk 7 above)
Comment (konekoya) [reply ohara-manager/client/src/components/Pipeline/Graph/Connector/ConnectorGraph.js]: You can remove this temp fix now. It's addressed in #3708 :)

**Pull request merged**